### PR TITLE
Add grpc transport

### PR DESCRIPTION
**Pull request by jonparrott**
@bjwatson @geigerj please review, as this will replace all of [`gax.auth`](https://github.com/googleapis/gax-python/blob/master/google/gax/auth.py) and most of [`gax.grpc`](https://github.com/googleapis/gax-python/blob/master/google/gax/grpc.py).

@nathanielmanistaatgoogle PTAL to verify grpc usage.

### CODE REVIEW

File: google/auth/transport/grpc.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        google_auth_credentials_metadata_plugin)
+
+    if not ssl_credentials:
+        ssl_credentials = grpc.ssl_channel_credentials()
+
+    # Combine the ssl credentials and the authorization credentials.
+    composite_credentials = grpc.composite_channel_credentials(
+        ssl_credentials, google_auth_credentials)
+
+    channel = grpc.secure_channel(target, composite_credentials)
```
Comment (nathanielmanistaatgoogle) [inline google/auth/transport/grpc.py]: Always replace
```
<name> = <expression>

return <name>
```
with
```
return <expression>
```
except in those few cases in which `<name>` tells the reader something really important that they don't already know.

File: tests/transport/test_grpc.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import mock
+
+
```
Comment (nathanielmanistaatgoogle) [inline tests/transport/test_grpc.py]: Too many blank lines here?

File: google/auth/transport/grpc.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    # Create the metadata plugin for inserting the authorization header.
+    google_auth_credentials_metadata_plugin = AuthMetadataPlugin(
+        credentials, request)
+
+    # Create a set of grpc.CallCredentials using the metadata plugin.
+    google_auth_credentials = grpc.metadata_call_credentials(
+        google_auth_credentials_metadata_plugin)
+
+    if not ssl_credentials:
```
Comment (nathanielmanistaatgoogle) [inline google/auth/transport/grpc.py]: `if ssl_credentials is None:`?

**Comment by jonparrott** [comment]
@nathanielmanistaatgoogle any idea why `grpc.AuthMetadataPlugin` doesn't exist on Python 2.7 but does for 3.4/3.5?

File: google/auth/transport/grpc.py (hunk 1 above)
Comment (jonparrott) [inline google/auth/transport/grpc.py]: Done.

File: tests/transport/test_grpc.py (hunk 2 above)
Comment (jonparrott) [inline tests/transport/test_grpc.py]: Yes, fixed.

File: docs/conf.py [hunk 4]
```diff
@@ -373,3 +373,4 @@
 # Autodoc config
 autoclass_content = 'both'
 autodoc_member_order = 'bysource'
+autodoc_mock_imports = ['grpc']
```
Comment (dhermes) [inline docs/conf.py]: TIL!

**Comment by jonparrott** [comment]
> @nathanielmanistaatgoogle any idea why grpc.AuthMetadataPlugin doesn't exist on Python 2.7 but does for 3.4/3.5?

It might actually be that grpc isn't being installed. Do you know why this would be the case?

File: pylintrc [hunk 5]
```diff
@@ -137,6 +137,12 @@ method-rgx=[a-z_][a-z0-9_]{2,40}$
 # especially those that implemented wordy RFCs.
 function-rgx=[a-z_][a-z0-9_]{2,40}$
 
+# Regular expression matching correct variable names
+# DEFAULT:  variable-rgx=[a-z_][a-z0-9_]{2,30}$
+# RATIONALE:  Prefer longer, more specific variable names to avoid ambiguity. 
+# especially for methods that implement wordy rfcs or use wordy libraries.
+variable-rgx=[a-z_][a-z0-9_]{2,40}$
```
Comment (dhermes) [inline pylintrc]: You could just use a shorter variable name

**Comment by nathanielmanistaatgoogle** [comment]
No idea why that might be happening.

**Comment by jonparrott** [comment]
> No idea why that might be happening.

Does gRPC support PyPy? Surely it must support 2.7, but it's failing there too.

**Comment by nathanielmanistaatgoogle** [comment]
PyPy support isn't completely formalized, but if it's not working for you we definitely want to know. You shouldn't have to disable gRPC Python when running with PyPy.

**Comment by jonparrott** [comment]
> PyPy support isn't completely formalized, but if it's not working for you we definitely want to know. You shouldn't have to disable gRPC Python when running with PyPy.

Do you have a tracking bug on gRPC for this?

**Comment by jonparrott** [comment]
@bjwatson @geigerj seems I've satisfied @dhermes and @nathanielmanistaatgoogle, can you please take a look?

**Comment by nathanielmanistaatgoogle** [comment]
@jonparrott: [issue 4221](https://github.com/grpc/grpc/issues/4221) tracks formal PyPy support; if you've found something not working I'm sure @thunderboltsid will want to hear about it.

File: tests/transport/test_grpc.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import mock
+
+import pytest
+
+try:
```
Comment (nathanielmanistaatgoogle) [inline tests/transport/test_grpc.py]: This strikes me as worth decorating with `TODO(<issue link>): unconditional import.` even if the linked issue is one outside this project. This `try:`/`except ImportError:` isn't something you want to be in place any longer than it has to be, right?

**Comment by jonparrott** [comment]
@nathanielmanistaatgoogle it just seems to not compile on Travis. I'll link over to the build on that issue.

**Comment by nathanielmanistaatgoogle** [comment]
If I'm reading the Travis output correctly, this passes on 2.7 now, right? Did you find any reason why it was failing yesterday? PyPy is now the only known problematic environment?

**Comment by jonparrott** [comment]
> If I'm reading the Travis output correctly, this passes on 2.7 now, right? 
Correct.

> Did you find any reason why it was failing yesterday?

Not sure, but I'm guessing it was the addition of `absolute_import` that fixed it.

> PyPy is now the only known problematic environment?

Correct.

File: tests/transport/test_grpc.py (hunk 6 above)
Comment (nathanielmanistaatgoogle) [inline tests/transport/test_grpc.py]: [`:-\`](https://github.com/google/oauth2client/pull/503#r61331868)

Well... okay.

Comment (dhermes) [inline tests/transport/test_grpc.py]: I'm also on team "No TODOs in source".

Comment (nathanielmanistaatgoogle) [inline tests/transport/test_grpc.py]: @dhermes: I thought I'd won you over! I've got your thumbs-up [right here](https://github.com/google/oauth2client/pull/503#r61331868)!

Comment (dhermes) [inline tests/transport/test_grpc.py]: No, you just got me to say "meh" and move on

Comment (soltanmm-google) [inline tests/transport/test_grpc.py]: (random uninvited two cents: 'no TODOs in code' makes sense for small projects where the expected number of issues to browse is small; e.g. you don't end up in the same pickle that people get into when they choose a singly-linked list over a doubly-linked list. Ostensibly Python isn't meant for large complicated projects and growing larger *isn't supposed to happen*, but, y'know,  ¯\\\_(ツ)_/¯ )

**Comment by thunderboltsid** [comment]
@jonparrott What version of PyPy are you using and on what environment? Also, what version of grpc are you using?

**Pull request merged**